service/src/java/org/apache/hive/service/servlet/LoginServlet.java

```diff
@@ -36,11 +36,17 @@ public LoginServlet(LdapAuthService ldapAuthService) {
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) {
-    ldapAuthService.authenticate(request, response);
-    RequestDispatcher dispatcher = request.getRequestDispatcher(HiveServer2.HS2_WEBUI_ROOT_URI);
+    boolean success = ldapAuthService.authenticate(request, response);
     PrintWriter out = null;
     try {
       out = response.getWriter();
+      final RequestDispatcher dispatcher;
+      if (success) {
+        dispatcher = request.getRequestDispatcher(HiveServer2.HS2_WEBUI_ROOT_URI);
+      } else {
+        request.setAttribute("login_failure_message", "Invalid username or password");
```

Comment:
`login_failure_message` is used in 2 places, it would be better to define it as a constant.

Reply:
done